### ISSUE TITLE
Remotely closed sockets on long running connections

### ISSUE THREAD

**Issue by adamcooke** [ostinelli/apnotic#45]
I've just started using Apnotic behind my [APNS Proxy](/adamcooke/apns-proxy) app (basically an HTTP endpoint that can receive notifications and send them over to Apple).

Under normal usage our worker works like this:

1. We make a connection to APNS.
2. We keep the connection object around and use it each time a notification needs to be sent.
3. If there's an error sending a notification on the connection, we catch the error, disconnect and make a new connection. 

Everything is mostly working very well except every few hours we get socket errors triggered outside of the normal flow.

```
rake aborted!
SocketError: Socket was remotely closed
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:104:in `rescue in block (2 levels) in ensure_open'
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:98:in `block (2 levels) in ensure_open'
EOFError: end of file reached
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:122:in `block in socket_loop'
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:119:in `loop'
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:119:in `socket_loop'
/app/vendor/bundle/ruby/2.3.0/gems/net-http2-0.14.1/lib/net-http2/client.rb:99:in `block (2 levels) in ensure_open'
Tasks: TOP => apns_proxy:worker
```

From looking at the source for HTTP2 & Apnotic, these seem to be raised within a thread that has `abort_on_exception` set to true. Whenever these errors occur, the whole process dies with no option for us to catch the error and handle it gracefully.

For APNS Proxy, we have a single worker which maintains multiple connections to Apple - one for each application that uses it so the whole worker dying because a single connection has failed isn't desirable. 

The actual worker code can be found here if you're interested: https://github.com/adamcooke/apns-proxy/blob/master/lib/apns_proxy/worker.rb

Any help would be appreciated.

**Comment by ostinelli** [comment]
From the README:

> In case that errors are encountered, Apnotic will raise the error and repair the underlying connection, but it will not retry the requests that have failed. This is by design, so that the job manager (Sidekiq, Resque,...) can retry the job that failed. For this reason, it is recommended to use a queue engine that will retry unsuccessful pushes.

See also #41.

**Comment by ostinelli** [comment]
Oops, closed by error.

You are correct, any error happening on the underlying socket will raise to the main thread (as you can see from [here](https://github.com/ostinelli/net-http2/blob/0.14.1/lib/net-http2/client.rb#L111)).

Since the client is obviously asynchronous, how would you go in raising socket errors to allow for a job repetition?

**Comment by adamcooke** [comment]
When the error occurs on the main thread, what about marking the client as `dead` and raising the error next time it is used to send data to the server? This would then mean the exception can be caught cleanly and that job can be retried rather than relying on the main thread to handle anything?

**Comment by adamcooke** [comment]
Having suggested that, it might not be ideal in a situation where you're waiting for data to be sent to you although that isn't really the case for Apnotic's use case.

The issue I've found with rescuing on the main thread is that I don't know which connection has actually failed so I would need to tear down all connections and start them all again which might not be needed. It would be nice if the SocketError could include the connection that has failed so it can be found later. Something like:

```ruby
begin
  # Do stuff as normal on main thread
rescue NetHttp2::SocketError => e
  e.client #=> NetHttp2::Client
end
```

This would then have an instance of `NetHttp2::Client` which can be used to identify which Apnotic connection instance needs burning. Although, for that to work, the `Apnotic::Connection` class would need to expose the underlying `@client` too.

The current `abort_on_exception` technique seems like it would cause many job workers to simply die and be re-spawned by a process manager rather than handling the issue themselves nicely. Obviously, they can already do this if they don't care which connection is having issues.

Ignoring the Apnotic thing for a moment, it would be nice for the `SocketError` to include the connection in the exception anyway.

**Comment by ostinelli** [comment]
> Personally, I'm not a fan of a program being able to raise errors outside of the expected flow of the program. In this case, you can "randomly" get an exception when you least expect it.

Fair enough, again all of this has been programmed and architected with a specific use-case in mind.

Are you willing to provide a PR on [NetHttp2](https://github.com/ostinelli/net-http2) for this?

**Comment by ostinelli** [comment]
Related to https://github.com/ostinelli/net-http2/issues/4

**Comment by adamcooke** [comment]
Sorry - just spotted this! I will submit a PR on NetHttp2 for at least one of these options.

**Comment by ostinelli** [comment]
@adamcooke no worries I'm currently in the process of writing this myself. Will come back to you to see if it will cover your concerns.

**Comment by ostinelli** [comment]
I've provided a possible patch, please provide feedback in https://github.com/ostinelli/net-http2/issues/4.

Closing here.

**Comment by ostinelli** [comment]
I released Apnotic `1.1.0`, can you please check and let me know if this suits your needs?
